### DISCUSSION TITLE
Repeating pipeline & global variable declaration

### DISCUSSION THREAD

**Discussion by PedroRiken** [bonsai-rx/bonsai#1405] [Q&A]
Hello!

Basically, this workflow is my first step at attempting to create a bonsai closed-loop system for stimuli presentation with [fictrac](https://github.com/rjdmoore/fictrac) tracking. Where the pipeline should repeat for a certain amount of times with a certain trial duration.

First and less important: I want the user to be able to assign some global variables related to the experimental session through bonsai editor and currently the whole workflow is full of these externalized variable properties. Is there "cleaner" way to do it? Something like grouping the variable nodes but still keeping their visibility at the editor property window.

![image](https://github.com/bonsai-rx/bonsai/assets/135926261/18d071f8-d22b-4648-9ea7-8911c31c1f99)

Then, I have been able to create a single trial pipeline, however, when I apply the repeatcount node (for the chosen number of trials) after trial completion it does not repeat... And I am not figuring out how to do it besides a periodic timer subscription. 

Moreover, I would also like to stop the workflow after the defined number of trials has been met and the trick of subscribing to an "Workflow output" node is not working here... 

![image](https://github.com/bonsai-rx/bonsai/assets/135926261/58b1fd60-9a88-4e5c-8d4d-0e599d6d4c33)

PS: I am quite to programming in bonsai, so if there are some optimization tips to the current nodes, it would be much appreciated.

Keep up the good work!

[CloseLoopStim_Courtship.zip](https://github.com/bonsai-rx/bonsai/files/11701702/CloseLoopStim_Courtship.zip)

**Comment by bruno-f-cruz** [comment]
You rarely want to use `Repeat` on branched workflows. Since the full branch does not close until all the subscribers are finished, the repeat will never trigger ([see here too](https://github.com/orgs/bonsai-rx/discussions/1328#discussioncomment-5688540)). Try to use it without having "hanging" nodes.
